api/src/opentrons/protocol_engine/state/commands.py

```
@@ -60,18 +72,43 @@ def get_all(self) -> List[Command]:
     def get_next_queued(self) -> Optional[str]:
         """Return the next request in line to be executed.
 
-        Normally, this corresponds to the earliest-added command that's currently
-        pending.
-
-        But if any command added before that command is currently failed, None is
-        returned instead. This models the entire protocol stopping when any command
-        fails.
-
-        If there are no pending commands at all, returns None.
+        Returns:
+            Normally, the ID of the earliest queued command. However, will return
+                None if any command has failed (since a command failure is fatal
+                to a protocol run) or if the engine is currently paused.
         """
+        if self._state.is_running is False:
+            return None
+
         for command_id, command in self._state.commands_by_id.items():
             if command.status == CommandStatus.FAILED:
                 return None
             elif command.status == CommandStatus.QUEUED:
                 return command_id
+
         return None
+
+    def get_is_complete(self, command_id: Optional[str] = None) -> bool:
+        """Get whether a given command is (or all commands are) completed.
+
+        A command is "completed" if one of the following is true:
+
+        - Its status is CommandStatus.SUCCEEDED
+        - Its status is CommandStatus.FAILED
+        - A command earlier in the queue has a status of CommandStatus.FAILED
+             - In this case, the command in question will never run
```

I find it a little unintuitive that this is called "completed." The whole protocol is complete, but that particular command never got a chance to run.

Is this behavior necessary to support calling code, or something?

---

I was imaging a call to `POST /commands/execute` (or something similar) when there are already commands running in the queue. If an earlier command failed and took the engine down (which is the current implementation) you wouldn't want your `POST` to just wait indefinitely I don't think

---

Hm, interesting.

At the HTTP level, what would we want that response to be? Would a 500 work? Because if so, we could just raise an exception, I think.

---

After discussion, leaving as is for now. Anticipating plenty of future work for failure recovery, so more work will go into this logic

---

Yea. The 'completed' response even if the command hasn't been executed, along with assigning a command status [`CommandStatus.SUCCEEDED` before executing](https://github.com/Opentrons/opentrons/pull/8152/files#diff-689985af568a93ee478d43b950c1af2c0bc54d409e900c35e755a434d5dd5be6R54) is definitely confusing. Cool with revisiting this later.